### PR TITLE
Save to local storage

### LINKED ISSUE #17: [TODO] Save current scene in localStorage
So we don't lose it if the tab is closed

### PR DESCRIPTION
**Pull request by paulomenezes**
- Save current canvas to local storage when something change
- Restore saved canvas when page load

Closes #17

### CODE REVIEW

File: src/index.tsx [hunk 1]
```diff
@@ -436,50 +463,62 @@ class App extends React.Component<{}, AppState> {
...
-                this.setState({ exportBackground: e.target.checked });
+        <div className="wrappers">
+          <div className="saveWrapper">
+            <button disabled={elements.length === 0} onClick={save}>
+              Save
+            </button>
+            <button disabled={!hasSavedItems} onClick={restore}>
+              Restore
+            </button>
+          </div>
```
Comment (vjeux) [inline src/index.tsx]: We should do it automatically whenever a change is done. And restore on page load.

File: yarn.lock [hunk 2]
```diff
@@ -1402,16 +1402,16 @@
   resolved "https://registry.yarnpkg.com/@types/q/-/q-1.5.2.tgz#690a1475b84f2a884fd07cd797c00f5f31356ea8"
   integrity sha512-ce5d3q03Ex0sy4R14722Rmt6MT07Ua+k4FwDfdcToYJcMKNtRVQvJ6JCAPdAmAnbRb6CsX6aYb9m96NGod9uTw==
 
-"@types/react-dom@^16.9.4":
+"@types/react-dom@16.9.4":
   version "16.9.4"
-  resolved "http://storage.mds.yandex.net/get-npm/1893413/@types/react-dom/-/react-dom-16.9.4.tgz#0b58df09a60961dcb77f62d4f1832427513420df"
```
Comment (vjeux) [inline yarn.lock]: This is weird that we're getting those from yandex.

Comment (paulomenezes) [reply yarn.lock]: Some modifications?

File: src/index.tsx [hunk 3]
```diff
@@ -358,6 +360,43 @@ function deleteSelectedElements() {
...
+  window.removeEventListener("beforeunload", onUnload);
+}
+
+function restore() {
+  const el = localStorage.getItem(LOCAL_STORAGE_KEY);
+
+  if (el) {
+    const items = JSON.parse(el);
+    for (let item of items) {
+      item = generateDraw(item, "#000000", "#ffffff");
```
Comment (vjeux) [inline src/index.tsx]: @Fausto95 I just realized but the stroke and fill colors should be a property of the elements, not a global one. We'll need to update this.

For example, right now you can create a black rectangle, change the color to red, copy paste the black rectangle which will turn red. This is unexpected.

File: src/index.tsx [hunk 4]
```diff
@@ -358,6 +360,43 @@ function deleteSelectedElements() {
...
+  }
+
+  window.removeEventListener("beforeunload", onUnload);
+}
+
+function restore() {
+  const el = localStorage.getItem(LOCAL_STORAGE_KEY);
+
+  if (el) {
+    const items = JSON.parse(el);
```
Comment (vjeux) [inline src/index.tsx]: We name them "elements" thorough the codebase, would be nice to use the same naming here.

File: src/index.tsx [hunk 5]
```diff
@@ -358,6 +360,43 @@ function deleteSelectedElements() {
...
+    for (let item of items) {
```
Comment (vjeux) [inline src/index.tsx]: So far we've been using .forEach instead of let. Would be nice to do the same if possible

File: src/index.tsx [hunk 6]
```diff
@@ -399,6 +439,7 @@ class App extends React.Component<{}, AppState> {
       drawScene();
       event.preventDefault();
     } else if (event.key === "Backspace") {
+      addOnBeforeUnload();
```
Comment (vjeux) [inline src/index.tsx]: This is very error prone to sprinkle those tests at places where we may be changing something. We should figure out the right place to hook this into.

Ideally, what we want is whenever the elements change, to save it. componentDidMount is likely a good candidate.

File: src/index.tsx [hunk 7]
```diff
@@ -408,6 +449,8 @@ class App extends React.Component<{}, AppState> {
       event.key === "ArrowUp" ||
       event.key === "ArrowDown"
     ) {
+      addOnBeforeUnload();
```
Comment (vjeux) [inline src/index.tsx]: If you know that something has changed, why not run the save instead of telling the user to do so?

File: src/index.tsx [hunk 8]
```diff
@@ -454,6 +497,57 @@ class App extends React.Component<{}, AppState> {
   public render() {
     return (
       <>
+        <div className="wrappers">
+          <div className="saveWrapper">
+            <button disabled={elements.length === 0} onClick={save}>
+              Save
```
Comment (vjeux) [inline src/index.tsx]: I think that an explicit save button isn't a good idea in practice here. I feel like we should just always save to localstorage and restore it when it comes back

File: src/index.tsx [hunk 9]
```diff
@@ -358,6 +360,43 @@ function deleteSelectedElements() {
   }
 }
 
+function save() {
+  if (elements && elements.length > 0) {
+    const items = [...elements];
+    for (const item of items) {
+      item.isSelected = false;
```
Comment (vjeux) [inline src/index.tsx]: I'm curious, why are you removing the selection? It may not be a bad idea to restore it as well.

File: src/index.tsx [hunk 10]
```diff
@@ -358,6 +360,43 @@ function deleteSelectedElements() {
   }
 }
 
+function save() {
+  if (elements && elements.length > 0) {
+    const items = [...elements];
```
Comment (vjeux) [inline src/index.tsx]: Note that this doesn't do a deep copy. So you're still mutating the original elements two lines below

File: src/index.tsx [hunk 11]
```diff
@@ -358,6 +360,43 @@ function deleteSelectedElements() {
   }
 }
 
+function save() {
+  if (elements && elements.length > 0) {
```
Comment (vjeux) [inline src/index.tsx]: why are you checking for length > 0? It seems like the logic below also works for 0 elements. Also, elements should never be null/undefined

File: src/index.tsx [hunk 12]
```diff
@@ -794,6 +892,8 @@ const context = canvas.getContext("2d")!;
 // https://stackoverflow.com/questions/13879322/drawing-a-1px-thick-line-in-canvas-creates-a-2px-thick-line/13879402#comment90766599_13879402
 context.translate(0.5, 0.5);
 
+restore();
```
Comment (vjeux) [inline src/index.tsx]: <3

File: src/index.tsx (hunk 4 above)
Comment (vjeux) [inline src/index.tsx]: You probably want to add a try catch here, we want to fail silently if the data is corrupted

**Comment by Fausto95** [comment]
Yes, that makes sense @vjeux

File: yarn.lock (hunk 2 above)
Comment (Fausto95) [reply yarn.lock]: I had this weird thing too, looks like it's a codesandbox thing

File: src/index.tsx (hunk 3 above)
Comment (Fausto95) [reply src/index.tsx]: Yes, sure! 
I can work on this as soon as I get some free time

File: yarn.lock (hunk 2 above)
Comment (nanot1m) [reply yarn.lock]: It's my fault. Forgot to change npm registry  😔

**Comment by paulomenezes** [comment]
@vjeux  I'll work on this suggestions

**Comment by paulomenezes** [comment]
How can we restore the state from local storage? I thought in the `restore` function return the state and then pass to App through props and set the state in the constructor, is it a good idea?

File: src/index.tsx (hunk 3 above)
Comment (Fausto95) [reply src/index.tsx]: Handled by https://github.com/excalidraw/excalibur/pull/65

File: src/index.tsx (hunk 12 above)
Comment (giovannigiordano) [reply src/index.tsx]: This can be moved to the `componentDidMount` and then update component's state

**Comment by giovannigiordano** [comment]
@paulomenezes I have added a comment, the restore function can be moved in the `componentDidMount` instead of constructor

File: src/index.tsx [hunk 13]
```diff
@@ -134,6 +137,8 @@ function renderScene(
   // null indicates transparent bg
   viewBackgroundColor: string | null
 ) {
+  if (!context) return;
```
Comment (paulomenezes) [inline src/index.tsx]: When I set the state in `componentDidMount`, the context is null

**Comment by vjeux** [comment]
Thanks so much for contributing, I just gave you access to the repo! Can't wait to see what cool stuff we're going to build together :)

**Comment by vjeux** [comment]
I've just been using it for the past half hour and it's so awesome to have the current scene persist across reloads and while developing!

**Comment by paulomenezes** [comment]
Is really nice!

**Pull request merged**